src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -150,8 +153,214 @@ CosineTree::CosineTree(const arma::mat& dataset,
   ConstructBasis(treeQueue);
 }
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+    // Copy matrix, but only if we are the root.
+    dataset((other.parent == NULL) ? new arma::mat(*other.dataset) : NULL),
+    delta(other.delta),
+    parent(NULL),
+    left(NULL),
+    right(NULL),
+    indices(other.indices),
+    l2NormsSquared(other.l2NormsSquared),
+    centroid(other.centroid),
+    basisVector(other.basisVector),
+    splitPointIndex(other.SplitPointIndex()),
+    numColumns(other.NumColumns()),
+    l2Error(other.L2Error()),
+    frobNormSquared(other.FrobNormSquared()),
+    localDataset(other.parent == NULL)
+{
+  // Create left and right children (if any).
+  if (other.Left())
+  {
+    left = new CosineTree(*other.Left());
+    left->Parent() = this; // Set parent to this, not other tree.
+  }
+
+  if (other.Right())
+  {
+    right = new CosineTree(*other.Right());
+    right->Parent() = this; // Set parent to this, not other tree.
+  }
+
+  // Propagate matrix, but only if we are the root.
+  if (parent == NULL && localDataset)
+  {
+    std::queue<CosineTree*> queue;
+    if (left)
+      queue.push(left);
+    if (right)
+      queue.push(right);
+    while (!queue.empty())
+    {
+      CosineTree* node = queue.front();
+      queue.pop();
+
+      node->dataset = dataset;
+      if (node->left)
+        queue.push(node->left);
+      if (node->right)
+        queue.push(node->right);
+    }
+  }
+}
+
+//! Copy assignment operator: copy the given other tree.
+CosineTree& CosineTree::operator=(const CosineTree& other)
+{
+  // Return if it's the same tree.
+  if (this == &other)
+    return *this;
+
+  // Freeing memory that will not be used anymore.
+  if (localDataset)
+    delete dataset;
+
+  delete left;
+  delete right;
+
+  // Performing a deep copy of the dataset.
+  dataset = new arma::mat(*other.dataset);
```

Comment:
So, either we should only allocate a new matrix here if we are the root, or always set `localDataset` to `true`. I think the first is the better approach.